### ISSUE TITLE
new code generation

### ISSUE THREAD

**Issue by drieks** [arrow-kt/arrow-meta#32]
Hello Arrow Team,

I started a new AST-Parsing Library `kotlinx.ast` (https://github.com/kotlinx/ast) that can be used to read Kotlin source files into easy to use data classes (using [antlr-kotlin](https://github.com/Strumenta/antlr-kotlin) and the [official Kotlin language grammar](https://kotlinlang.org/docs/reference/grammar.html)).

I'm already using this library for code generation, in combination with `kotlinpoet` (https://github.com/square/kotlinpoet) to create the generated files.

Maybe it can be used in `arrow-meta`. Please let me know your requirements for code generation if you are interested in discussing this.

`kotlinx.ast` is currently JVM-only, but support for JavaScript and Native is planned. It should be easy to create a shared code generation module for `arrow-meta` that can be used inside a gradle plugin, in unit tests and even in a native code generation tool (using kotlin native).

@rachelcarmena , @raulraja: A starting point for discussing may be replacing `kastree` with `kotlinx.ast` (https://github.com/arrow-kt/arrow/tree/rr-meta-prototype-integration/modules/meta/arrow-meta-prototype/compiler-plugin/src/main/kotlin/kastree/ast)

**Comment by raulraja** [comment]
Makes sense. Yes a feature like this would be great, we would need then figure out what needs to be changed in the IDE part to treat those as valid PSI elements

**Comment by LifeIsStrange** [comment]
Are you aware that Kotlin 1.4 is changing its intermediary representation? (in order to unify it for cross platform)
Keywords: Klib, FIR 
https://blog.jetbrains.com/kotlin/2019/12/what-to-expect-in-kotlin-1-4-and-beyond/
